astropy/utils/tests/test_data.py

```diff
@@ -645,6 +647,28 @@ def test_download_cache():
     assert not os.path.isdir(lockdir), "Cache dir lock was not released!"
 
 
+@pytest.mark.remote_data(source="astropy")
+def test_download_certificate_verification_failed():
+    """Tests for https://github.com/astropy/astropy/pull/10434"""
+
+    # First test the expected exception when download fails due to a
+    # certificate verification error; we simulate this by passing a bogus
+    # CA directory to the ssl_context argument
+    ssl_context = {'cafile': None, 'capath': '/does/not/exist'}
+    with pytest.raises(urllib.error.URLError) as exc:
+        download_file(TESTURL_SSL, cache=False, ssl_context=ssl_context)
+
+    msg = f'verification of TLS/SSL certificate at {TESTURL_SSL} failed'
+    assert msg in str(exc.value)
+
+    with pytest.warns(AstropyWarning) as warning_lines:
+        fnout = download_file(TESTURL_SSL, cache=False,
+                ssl_context=ssl_context, allow_insecure=True)
+
+    assert os.path.isfile(fnout)
+    assert len(warning_lines) == 1 and msg in str(warning_lines[0])
```

Comment:
Message check moved to `match=` above but still need to make sure this is the only warning emitted.
```suggestion
    assert len(warning_lines) == 1
```